common/kvstore/src/main/java/org/apache/spark/util/kvstore/InMemoryStore.java

```diff
@@ -205,23 +211,48 @@ public void accept(Comparable<Object> key, T value) {
     private final KVTypeInfo ti;
     private final KVTypeInfo.Accessor naturalKey;
     private final ConcurrentMap<Comparable<Object>, T> data;
+    private final String naturalParentIndexName;
+    // A mapping from parent to the natural keys of its children.
+    // For example, a mapping from a stage ID to all the task IDs in the stage.
+    private final ConcurrentMap<Comparable<Object>, NaturalKeys> parentToChildrenMap;
 
     private InstanceList(Class<?> klass) {
       this.ti = new KVTypeInfo(klass);
       this.naturalKey = ti.getAccessor(KVIndex.NATURAL_INDEX_NAME);
       this.data = new ConcurrentHashMap<>();
+      this.naturalParentIndexName = ti.getParentIndexName(KVIndex.NATURAL_INDEX_NAME);
+      this.parentToChildrenMap = new ConcurrentHashMap<>();
     }
 
     KVTypeInfo.Accessor getIndexAccessor(String indexName) {
       return ti.getAccessor(indexName);
     }
 
     int countingRemoveAllByIndexValues(String index, Collection<?> indexValues) {
```

Comment:
since there is no document, I still have problems understanding it.

The basic idea is: if the `index` is natural index, we can just look it up in O(1). For other index, we have to do linear scan, extract keys and find matches.

The extension here: if the `index` is parent of natural index, get the children natural indexes and do O(1) lookup.

However, seems the basic idea is missing? Shall we avoid linear scan if `index` is natural index?

Reply:
Yes and no.
I am aware of the natural index is not handled properly here. But all the method calls of this method doesn't pass the natural index in. 
There is a simpler API to use in `KVStore`
```
void delete(Class<?> type, Object naturalKey)
```

So I think this is minor and I prefer to do it in another PR, since this one is complicated.

Reply:
we should probably unify the `delete` method, but it's not related to this PR.

Reply:
Yes,  I will do it in another PR.